src/test/java/spoon/test/snippets/SnippetTest.java

```diff
@@ -239,4 +242,15 @@ public void testSnippetCompilationInUnnamedPackage() {
 		assertTrue(body.getStatements().get(0) instanceof CtLocalVariable);
 		assertEquals(1,body.getStatements().size()); 
 	}
+
+	@Test
+	public void testSnippetsWithDifferingValuesAreNotEqual() {
+		// contract: Two snippets with different values should be considered not equal
+		Factory factory = new Launcher().getFactory();
+
+		CtExpression<Integer> one = factory.createCodeSnippetExpression("1");
+		CtExpression<Integer> two = factory.createCodeSnippetExpression("2");
+
+		assertThat(one, not(equalTo(two)));
```

Comment:
Any reason why you have preferred `asserThat` over `assertEquals`?

Reply:
> Any reason why you have preferred asserThat over assertEquals?

what about `assertNotEquals(one, two);` 😅

Reply:
> Any reason why you have preferred asserThat over assertEquals?

Many reasons. First, I prefer the way you write the Hamcrest matchers, i.e. `assertThat(one, not(equalTo(expected)))`. It's clear which is the actual value, and which is the expected one. For `assertNotEquals(one, two)`, not so much (these ar actually reversed, because it's `assertNotEquals(expected, actual), but `assertThat(actual, someMatcher(expected))`).

Hamcerst matchers also almost always give more informative failure information than JUnit's built-in matchers when an assertion fails.

There's however a pretty big problem with this test that is unrelated to the assertion being used. Why is `one` the actual value, and `two` the expected value? Which method is actually under test here? Of course, neither of `one` or `two` is actual or expected, they are just two values. The method under test is `equals`, but that's hidden by the assertion.

The test should probably look like this instead:

```suggestion
		boolean areEqual = one.equals(two);

		assertFalse(areEqual);
```

Reply:
Or more concisely, this would also be fine
```suggestion
		assertFalse(one.equals(two));
```

Asserting true and false is probably the one place where I prefer the built-in assertions to Hamcrest matchers, because the Hamcrest alternative is a bit more verbose than necessary:

```suggestion
		assertThat(one.equals(two), is(false));
```

However, compare the Hamcrest failure message:

```
java.lang.AssertionError: 
Expected: is <false>
     but: was <true>
[... stacktrace ...]
```

Is much more readable than the JUnit failure message:

```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:87)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertFalse(Assert.java:65)
	at org.junit.Assert.assertFalse(Assert.java:75)
	at spoon.test.snippets.SnippetTest.testSnippetsWithDifferingValuesAreNotEqual(SnippetTest.java:255)
[... more stacktrace ...]
```

So actually I'll probably go with the Hamcrest one after all :)

Reply:
> java.lang.AssertionError: 
Expected: is <false>
     but: was <true>
[... stacktrace ...]

Agree, this is clearly better.